File: Python/Algorithm I/733. Flood Fill.py
def flood_fill(image: list[list[int]], start_row: int, start_column: int, color: int) -> list[list[int]]:
    
    MAX_ROW = len(image)
    MAX_COLUMN = len(image[0])

    old_color = image[start_row][start_column]

    def fill(row: int, column: int) -> None:
        
        if not (0 <= row < MAX_ROW and 0 <= column < MAX_COLUMN):
            return

        if image[row][column] != old_color or image[row][column] == color:
            return

        image[row][column] = color

        fill(row - 1, column)
        fill(row + 1, column)
        fill(row, column - 1)
        fill(row, column + 1)

    fill(start_row, start_column)

    return image

File: TypeScript/Algorithm/Algorithm I/733. Flood Fill.ts
/**
 * Modifies the image by changing the color of connected cells starting from a given cell.
 * @param {number[][]} image - The image represented as a matrix of colors.
 * @param {number} sr - The row index of the starting cell.
 * @param {number} sc - The column index of the starting cell.
 * @param {number} color - The new color to be applied.
 * @returns {number[][]} The modified image with the updated color.
 */
function floodFill(image: number[][], sr: number, sc: number, color: number): number[][] {
    const ROWS: number = image.length;
    const COLS: number = image[0].length;

    const oldColor: number = image[sr][sc];

    if (oldColor === color)
        return image;

    function dfs(row: number, col: number): void {
        if (row < 0 || col < 0 || row >= ROWS || col >= COLS || image[row][col] !== oldColor)
            return;

        image[row][col] = color;

        dfs(row - 1, col); // Up
        dfs(row + 1, col); // Down
        dfs(row, col - 1); // Left
        dfs(row, col + 1); // Right
    }

    dfs(sr, sc);

    return image;
};
